nav2_costmap_2d/src/collision_checker.cpp

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) 2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_costmap_2d/collision_checker.hpp"
+#include "nav2_costmap_2d/cost_values.hpp"
+#include "nav2_util/line_iterator.hpp"
+#include "nav2_costmap_2d/exceptions.hpp"
+#include "nav2_costmap_2d/footprint.hpp"
+
+namespace nav2_costmap_2d
+{
+
+CollisionChecker::CollisionChecker(
+  rclcpp::Node::SharedPtr ros_node,
+  std::shared_ptr<CostmapSubscriber> costmap_sub,
+  std::shared_ptr<FootprintSubscriber> footprint_sub,
+  tf2_ros::Buffer & tf_buffer,
+  std::string name)
+: tf_buffer_(tf_buffer),
+  node_(ros_node), name_(name),
+  costmap_sub_(costmap_sub), footprint_sub_(footprint_sub)
+{
+  node_->get_parameter_or<std::string>(
+    "global_frame", global_frame_, std::string("map"));
+  node_->get_parameter_or<std::string>(
+    "robot_base_frame", robot_base_frame_, std::string("base_link"));
+}
+
+CollisionChecker::~CollisionChecker() {}
+
+bool CollisionChecker::isCollisionFree(
```

Comment:
I'll look and give some thought into returning enum types. The exceptions here are modified from those in DWB as you mentioned.

Reply:
IMO exceptions should be used when the local code can't recover from the error but someone up the stack may be able to. They shouldn't be used to propagate a normal return value. For example, if isCollisionFree may commonly return true or false, a return value is more appropriate. So, both exceptions and return values are appropriate, depending on the circumstances, and should be applied the right way.

Reply:
I could modify `scorePose` to return a boolean and replace  `isCollisionFree`, but in this PR I was moving the code (including the exceptions) from the `ObstacleFootprint` critic which returns a score for the pose requested. My intention was to replace the `ObstacleFootprint` critic code with this class to avoid code duplication, which is why I thought to keep it as is.